cluster-autoscaler/proposals/pricing.md

```diff
@@ -0,0 +1,177 @@
+ 
+# Choosing the best node pool to expand
+ 
+Knowing the cost of a single node is only a part of the story. We need to choose the best node pool to accommodate the unscheduled pods, as a whole. However putting all of the pending pods to a single node pool may not be always an option because:
+ 
+* Node pool min/max boundaries. Pending pods may require expanding it beyond the max range.
+* Some pods may not fit into the machines.
+* Different node selectors.
+ 
+Different node pools may accept different pods and accommodate them on different number of nodes what can result in completely different cost. Let's denote the costs of the expansion as C.
+
+Let me give you an example.
+ 
+* Option1:  requires 3 nodes of Type1, accommodates pods P1, P2, P3 and costs C1=10$
+* Option2:  requires 2 nodes of Type2, accommodates pods P1, P3, P4, P5 and costs C2=20$. 
+ 
+It is hard to tell whether we will get a better deal with paying 10$ for having 3 pods running or with paying 20$ for a different set of pods. We need to make C1 and C2 somehow comparable.
+ 
+We can compute how much would it cost to run a pod on a machine that is perfectly suited to its needs. From GKE pricing we know that:
+ 
+ * 1 cpu cost $0.033174 / hour,  
+ * 1 gb of memory cost $0.004446 / hour
+ * 1 GPU is 0.7 / hour. 
+ 
+We can simplify pod ssd storage request and assume that it needs 50gb for local storage that cost 0.01$ per hour. 
+ 
+For two expansion options we could compute what is the theoretical cost of having all of these pods running on perfectly-fitted machines and mark it as T1 and T2. 
+ 
+Then C1/T1 and C2/T2 would denote how effective are we with a particular node group expansion. For example C1/T1 may equal to 2 which means that we pay 2 times more than we would need in a perfect world. And if C2/T2 equal to 1.05 it means that we are super effective and it’s hard to expect that some other option would be much better. 
+ 
+If we consistently pick the option with the lowest real to theoretical cost we should get quite good approximation of the perfect (from the money perspective) cluster layout. It may not be the most optimal but finding the most optimal one seems to be an NP problem (it’s a kind of binpacking).  
+ 
+ 
+# Adding “preferred” node type to the formula
+ 
+C/T is a simple formula that doesn’t include other aspects as, for example, the preference to have bigger machines in a bigger cluster or have more consistent set of nodes. The advantage of big nodes is that they usually offer smaller resource fragmentation and are more likely to have enough resources to accept new pods. For example 2 x N1-standard-2 packed to 75% will only accept pods requesting less than 0.5 cpu, while 1 x N1-standard-2 can take 2 pods requesting 0.5 cpu OR 1 pod requesting
+1 cpu. Having more consistent set of machines makes cluster management easier.  
+ 
+To include this preference in the formula we introduce a per-node metric called NodeUnFitness. 
```

Comment:
Interested to see how big of a difference NodeUnfitness would make in pricing vs just using the `least-waste`expander. Seems like it would be almost negligible for most cluster setups?